File: app/src/main/java/com/sber/rupassword/MainActivity.java
package com.sber.rupassword;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends Activity implements CompoundButton.OnCheckedChangeListener {

    PasswordHelper helper;

    private EditText mPasswordInput;
    private TextView mResultText;
    private TextView mResultGeneratedText;
    private TextView mPasswordLength;
    private TextView mPasswordInputBits;
    private TextView mPasswordGeneratedBits;
    private ImageButton mCopyButton;
    private ImageButton mCopyButtonGenerated;
    private ImageButton mRegeneratePassword;
    private ImageView mComplexityInput;
    private ImageView mComplexityGenerated;
    private SeekBar mSeekBar;
    private CompoundButton mCaps;
    private CompoundButton mSymbols;
    private CompoundButton mNumbers;

    private String[] mRussians;
    private String[] mLatins;

    private int MIN_QUANTITY = 4;
    private int mSymbolsQuantity = MIN_QUANTITY;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_new);

        mRussians = getResources().getStringArray(R.array.russians);
        mLatins = getResources().getStringArray(R.array.latins);
        helper = new PasswordHelper(mRussians, mLatins);

        mPasswordInput = findViewById(R.id.password_input);
        mResultText = findViewById(R.id.text_result);
        mResultGeneratedText = findViewById(R.id.text_result_generated);
        mPasswordLength = findViewById(R.id.password_length);
        mPasswordInputBits = findViewById(R.id.password_input_bits);
        mPasswordGeneratedBits = findViewById(R.id.password_generated_bits);
        mCopyButton = findViewById(R.id.copy_password_button);
        mCopyButtonGenerated = findViewById(R.id.copy_password_generated_button);
        mRegeneratePassword = findViewById(R.id.regenerate_password_button);
        mComplexityInput = findViewById(R.id.complexity_input);
        mComplexityGenerated = findViewById(R.id.complexity_generated);
        mSeekBar = findViewById(R.id.seekbar);
        mCaps = findViewById(R.id.check_caps);
        mSymbols = findViewById(R.id.check_symbols);
        mNumbers = findViewById(R.id.check_numbers);

        mCaps.setOnCheckedChangeListener(this);
        mSymbols.setOnCheckedChangeListener(this);
        mNumbers.setOnCheckedChangeListener(this);

        mCopyButton.setEnabled(false);
        mCopyButton.setOnClickListener(view -> copyToClipboard(mResultText));
        mCopyButtonGenerated.setOnClickListener(view -> copyToClipboard(mResultGeneratedText));
        mRegeneratePassword.setOnClickListener(view -> generatePassword());

        setTextViews();
        generatePassword();

        mPasswordInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                mResultText.setText(helper.convert(s));
                mCopyButton.setEnabled(s.length() > 0);

                int bits = helper.calculateStrength(mResultText.getText().toString());
                mPasswordInputBits.setText(getString(R.string.bits, bits));
                mComplexityInput.getDrawable().setLevel(bits);
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });

        mSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                mSymbolsQuantity = MIN_QUANTITY + i;
                setTextViews();
                generatePassword();
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

    }

    @Override
    public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
        generatePassword();
    }

    private void generatePassword() {
        mResultGeneratedText.setText(helper.generatePassword(mSymbolsQuantity, mCaps.isChecked(),
                mSymbols.isChecked(), mNumbers.isChecked()));

        int bits = helper.calculateStrength(mResultGeneratedText.getText().toString());
        mPasswordGeneratedBits.setText(getString(R.string.bits, bits));
        mComplexityGenerated.getDrawable().setLevel(bits);
    }

    private void copyToClipboard(TextView password) {
        ClipboardManager manager = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        if (manager != null) {
            manager.setPrimaryClip(ClipData.newPlainText(
                    getString(R.string.password), password.getText().toString()));
            Toast.makeText(MainActivity.this, R.string.toast_text_copied, Toast.LENGTH_SHORT).show();
        }
    }

    private void setTextViews() {
        String symbols = getResources().getQuantityString(
                R.plurals.symbols_count, mSymbolsQuantity, mSymbolsQuantity);
        mPasswordLength.setText(symbols);
        mPasswordInputBits.setText(getString(R.string.bits, 0));
    }
}

File: app/src/main/java/com/sber/rupassword/MainActivity.kt
package com.sber.rupassword

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.CompoundButton
import android.widget.SeekBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.doOnTextChanged
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), CompoundButton.OnCheckedChangeListener {

    private lateinit var passwordHelper: PasswordHelper

    private lateinit var ruLetters: Array<String>
    private lateinit var enLetters: Array<String>

    private var symbolsQuantity = MIN_QUANTITY

    private lateinit var capsBtn: CompoundButton
    private lateinit var symbolsBtn: CompoundButton
    private lateinit var numbersBtn: CompoundButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(findViewById(R.id.toolbar1))

        ruLetters = resources.getStringArray(R.array.russians)
        enLetters = resources.getStringArray(R.array.latins)

        passwordHelper = PasswordHelper(ruLetters, enLetters)

        capsBtn = findViewById(R.id.check_caps)
        symbolsBtn = findViewById(R.id.check_symbols)
        numbersBtn = findViewById(R.id.check_numbers)

        check_caps.setOnCheckedChangeListener(this)
        check_symbols.setOnCheckedChangeListener(this)
        check_numbers.setOnCheckedChangeListener(this)

        copy_password.isEnabled = false

        regenerate_password.setOnClickListener {
            generatePassword()
        }
        copy_password.setOnClickListener {
            copyToClipboard(password_result)
        }
        copy_password_generated.setOnClickListener {
            copyToClipboard(password_generated_result)
        }

        password_bits.text = getString(R.string.bits, 0)

        setPlurals()
        generatePassword()

        password_input.doOnTextChanged { text, _, _, _ ->
            password_result.text = passwordHelper.convert(text)
            copy_password.isEnabled = text?.length != 0

            val bits: Int = passwordHelper.calculateStrength(password_result.text.toString())
            password_bits.text = getString(R.string.bits, bits)
            password_complexity.drawable.level = bits
        }

        password_length_seekbar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                symbolsQuantity = MIN_QUANTITY + progress
                setPlurals()
                generatePassword()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
                //leave this empty
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                //leave this empty
            }

        })

    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
            when (item.itemId) {
                R.id.action_show_saved_passwords -> {
                    startActivity(Intent(this, SavedPasswordsActivity::class.java))
                    true
                }
                else -> super.onOptionsItemSelected(item)
            }

    override fun onCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
        generatePassword()
    }

    private fun generatePassword() {
        password_generated_result.text = passwordHelper.generatePassword(symbolsQuantity,
                check_caps.isChecked, check_symbols.isChecked, check_numbers.isChecked)

        val bits = passwordHelper.calculateStrength(password_generated_result.text.toString())
        password_generated_bits.text = getString(R.string.bits, bits)
        password_generated_complexity.drawable.level = bits
    }

    private fun setPlurals() {
        val symbols: String = resources.getQuantityString(R.plurals.symbols_count, symbolsQuantity,
                symbolsQuantity)
        password_length?.text = symbols
    }

    private fun copyToClipboard(password: TextView) {
        val manager: ClipboardManager =
                getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        manager.let {
            it.setPrimaryClip(
                    ClipData.newPlainText(getString(R.string.password), password.text.toString()))
            Toast.makeText(this, R.string.toast_text_copied, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private const val MIN_QUANTITY = 4
    }
}